src/transformers/models/bark/modeling_bark.py

```diff
@@ -0,0 +1,1372 @@
+        super().__init__()
+        self.weight = nn.Parameter(torch.ones(ndim))
+        self.bias = nn.Parameter(torch.zeros(ndim)) if bias else None
+
+    def forward(self, input):
+        return F.layer_norm(input, self.weight.shape, self.weight, self.bias, 1e-5)
+
+
+class MLP(nn.Module):
+    def __init__(self, config):
+        super().__init__()
+        self.in_proj = nn.Linear(config.hidden_size, 4 * config.hidden_size, bias=config.bias)
+        self.out_proj = nn.Linear(4 * config.hidden_size, config.hidden_size, bias=config.bias)
+        self.dropout = nn.Dropout(config.dropout)
+        self.gelu = nn.GELU()
+
+    def forward(self, x):
+        x = self.in_proj(x)
+        x = self.gelu(x)
+        x = self.out_proj(x)
+        x = self.dropout(x)
+        return x
+
+
+class BarkBlock(nn.Module):
+    def __init__(self, config, layer_idx, is_causal=False):
+        super().__init__()
+
+        if is_causal:
+            # if causal, uses handmade LayerNorm, so that the layerNorm bias is optional
+            # this handmade layerNorm is used to stick with Bark choice of leaving optional bias in AutoRegressive models
+            # (corresponding to the "Text" and the "Coarse" modules)
+            self.ln_1 = LayerNorm(config.hidden_size, bias=config.bias)
+        else:
+            self.ln_1 = nn.LayerNorm(config.hidden_size)
+
+        self.attn = BarkSelfAttention(config, is_causal=is_causal)
+
+        if is_causal:
+            self.ln_2 = LayerNorm(config.hidden_size, bias=config.bias)
+        else:
+            self.ln_2 = nn.LayerNorm(config.hidden_size)
+
+        self.mlp = MLP(config)
+        self.layer_idx = layer_idx
+
+    def forward(
+        self,
+        hidden_states,
+        past_kv=None,
+        attention_mask=None,
+        head_mask=None,
+        use_cache=False,
+        output_attentions=False,
+    ):
+        attn_outputs = self.attn(
+            self.ln_1(hidden_states),
+            past_kv=past_kv,
+            attention_mask=attention_mask,
+            head_mask=head_mask,
+            use_cache=use_cache,
+            output_attentions=output_attentions,
+        )
+
+        attn_output = attn_outputs[0]  # output_attn: output, present_kv, (attn_weights)
+        outputs = attn_outputs[1:]
+
+        hidden_states = hidden_states + attn_output
+        hidden_states = hidden_states + self.mlp(self.ln_2(hidden_states))
+
+        if use_cache:
+            outputs = (hidden_states,) + outputs
+        else:
+            outputs = (hidden_states,) + outputs[1:]
+
+        return outputs  # hidden_states, ((present), attentions)
+
+
+class BarkPreTrainedModel(PreTrainedModel):
+    """
+    An abstract class to handle weights initialization and a simple interface for downloading and loading pretrained
+    models.
+    """
+
+    config_class = BarkConfig
+    base_model_prefix = "bark"
+    supports_gradient_checkpointing = False
+
+    def _init_weights(self, module):
```

Comment:
We should probably initialise the linear, embedding and layer norm weights (custom layer norm that is!) according to the standard deviation we define for weight init: https://github.com/huggingface/transformers/blob/4124a09f8b3349f338917ad3282ca952bd15ec3a/src/transformers/models/gpt_neo/modeling_gpt_neo.py#L371

Reply:
I am not sure to understand this, can you explain a bit more ?

Reply:
Hmm I see that we actually define two PreTrainedModels:
1. BarkPreTrainedModel: https://github.com/huggingface/transformers/blob/2096d98af1855dc73b2a58611d4c8df69d7965d7/src/transformers/models/bark/modeling_bark.py#L369
2. BarkModulePreTrainedModel: https://github.com/huggingface/transformers/blob/2096d98af1855dc73b2a58611d4c8df69d7965d7/src/transformers/models/bark/modeling_bark.py#L387

=> would it work just having one pre-trained model here? Think this would simplify the code and class logic. The only thing that we need to do is define the config attribute `initializer_range` in the parent config (`config. initializer_range`) as well as the sub-model configs (e.g. `config.text_model. initializer_range`)

Reply:
I do it already anyways for `BarkFineModel`, so it should be ok

Reply:
Yep I think we should fuse the two pre-trained models and override the config class as you are doing 👍